### PR TITLE
overlord/config: normalize nulls to support config unsetting semantics

### PR DESCRIPTION
**Pull request by stolowski**
Purge (normalize) null values internally in snap configuration to support `snap unset` (`snap set foo!`) semantics (not implemented yet - will be done in a followup). Setting values to null with `snap set mysnap foo=null` will now remove the value from configuration. Normalization is done before commit and also after applying temporary changes on get.

### CODE REVIEW

File: overlord/configstate/config/helpers.go [hunk 1]
```diff
@@ -70,6 +110,7 @@ func PatchConfig(snapName string, subkeys []string, pos int, config interface{},
 		if err != nil {
 			return nil, err
 		}
+
```
Comment (pedronis) [inline overlord/configstate/config/helpers.go]: intentional?

File: overlord/configstate/config/helpers.go [hunk 2]
```diff
@@ -47,6 +47,46 @@ func ParseKey(key string) (subkeys []string, err error) {
...
+	case map[string]*json.RawMessage:
+		for k, v := range config {
+			cfg := purgeNulls(v)
+			if cfg != nil {
+				config[k] = cfg.(*json.RawMessage)
+			} else {
+				delete(config, k)
+			}
+		}
+		// After removing nulls from the map we can end up with an empty map
```
Comment (pedronis) [inline overlord/configstate/config/helpers.go]: this wasn't part of the plan, the plan mentions `null`,  not empty maps

File: overlord/configstate/config/helpers.go [hunk 3]
```diff
@@ -47,6 +47,46 @@ func ParseKey(key string) (subkeys []string, err error) {
...
+			return nil
+		}
+	case map[string]interface{}:
+		for k, v := range config {
+			if purgeNulls(v) == nil {
+				delete(config, k)
+			}
+		}
+		// After removing nulls from the map we can end up with an empty map
+		if len(config) == 0 {
```
Comment (pedronis) [inline overlord/configstate/config/helpers.go]: same

File: overlord/configstate/config/transaction.go [hunk 4]
```diff
@@ -172,6 +172,14 @@ func (t *Transaction) Get(snapName, key string, result interface{}) error {
 	// commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
 	config := t.copyPristine(snapName)
 	applyChanges(config, t.changes[snapName])
+
+	out := purgeNulls(config)
+	if out == nil {
```
Comment (pedronis) [inline overlord/configstate/config/transaction.go]: would this happen if we don't normalize away empty maps?

File: overlord/configstate/config/helpers.go (hunk 2 above)
Comment (stolowski) [reply overlord/configstate/config/helpers.go]: Right; I didn't think of them when we discussed the plan, and I wasn't sure if we considered this. When implementing this I found leaving empty maps around slightly confusing. I can drop it.

File: overlord/configstate/config/transaction.go [hunk 5]
```diff
@@ -172,6 +172,14 @@ func (t *Transaction) Get(snapName, key string, result interface{}) error {
 	// commit changes onto a copy of pristine configuration, so that get has a complete view of the config.
 	config := t.copyPristine(snapName)
 	applyChanges(config, t.changes[snapName])
+
+	out := purgeNulls(config)
```
Comment (bboozzoo) [inline overlord/configstate/config/transaction.go]: Wouldn't it still work as before if you have this:
```
purgeNulls(config)
return getFromConfig(snapName, subkeys, 0, config, result)
```

Comment (stolowski) [reply overlord/configstate/config/transaction.go]: Yes, sure, note I'm not passing `out` to getFromConfig and the only reason for special casing this condition is to higlight the fact that that entire configuration was "reduced" to `nil` (by removing empty maps) while still in the transaction. We return `NoOptionError` in such case while maybe we should give 'snap has no configuration` error, so this special casing here would make handling it easy and obvious if we wanted to (but if we do that then we would need to change implementation of `GetFeatureFlag` which wants to report true or false if there is no config).

Nb, this special-case goes away with the change to not reduce empty maps per @pedronis comments.

File: overlord/configstate/config/transaction.go (hunk 4 above)
Comment (stolowski) [reply overlord/configstate/config/transaction.go]: We return `NoOptionError` regardless of normalizing empty maps, the idea for this special case here was to make it easier to transform it into 'snap has no configuration'  error if we wanted to, see also the above explanation to @bboozzoo's question. Removing it for now as with the removal of empty maps normalization such case would need different conditonals.

File: overlord/configstate/config/helpers.go (hunk 2 above)
Comment (stolowski) [reply overlord/configstate/config/helpers.go]: Removed.

File: overlord/configstate/config/helpers.go (hunk 3 above)
Comment (stolowski) [reply overlord/configstate/config/helpers.go]: Removed.

**Comment by stolowski** [comment]
> lgtm, the commit/description should clarify that this makes already `snap set mysnap foo=null` remove `foo`. afaict there's already a spread test here about that.

Thanks. Commit message adjusted. And yes, the `snap-set` spread test about that.

**Pull request merged**